scripts/9 Rust on Rails.md

```diff
@@ -0,0 +1,539 @@
+- or an error 
+
+Now the compiler is happy.
+And if the compiler is happy, I am happy.
+
+We have clearly stated in the function signature that you won't get an integer back from this function, you will get it wrapped in a Result and you must handle it.
+
+Treating Errors as just another value is one of the features of Rust that allow you to fearlessly write complex distributed systems confidently.
+
+An example of a company building large-scale complex distributed systems in Rust is today's sponsor Ditto.
+
+
+---
+
+
+# ![[scripts/private/ditto_logo_white_text.png]]
+
+notes:
+
+- Unlike other sponsors, ditto don't want your money, they actually want to pay you.
+- This is because they've asked me to tell you about open Rust positions at their company.
+- Here's what you'll want to know about them, their tech, and their open positions:
+
+---
+
+### Ditto mesh architecture 
+![[ditto_mesh_architecture.png|850]]
+notes:
+
+- Ditto use rust to power their cross-platform data sync system
+- They're growing their team and are looking for people passionate about Rust, if you're watching this video, that might be you.
+- The problems they are solving include mesh networking, replication protocols, Conflict-free Replicated Data Types, and database design to name a few.
+
+---
+### Ditto rainbow connection
+![[ditto_rainbow connection.png]]
+
+notes:
+
+
+- They're looking for people with
+    - demonstrable Rust experience or previous work with C/C++
+    - Rust backend developers for their Big Peer cloud system
+    - Low-level bare-metal coders working with FFI and C interop
+    - Algorithm junkies to work on their data stores, and
+    - Networking coders at either the low or high level to work on their actor-like frameworks in replication and multihop work.
+
+---
+
+![[scripts/private/ditto_logo_white_text.png]]
+
+## [ditto.live](https://www.ditto.live) 
+
+[ditto.live/jobs](http://www.ditto.live/jobs)
+
+notes:
+
+Find out more about ditto at ditto.live, and see their
+open positions at ditto.live/jobs.
+
+My thanks to ditto for their support of my work.
+
+Let's get back on the rails:
+
+---
+
+
+## Railway-oriented programming
+
+![[rails-1-success-failure.png]]
+
+
+[fsharpforfunandprofit.com/rop/](https://fsharpforfunandprofit.com/rop/)
+
+&mdash; [@ScottWlaschin](https://twitter.com/ScottWlaschin)
+
+notes:
+
+- You may have heard of this metaphor, It's one I found very useful when trying to understand functional errors like the Result type in Rust. 
+- It was created by, and I will be borrowing a few slides from, Scott Wlaschin (pronounced: Vuh-losh-in) in 2014
+- Each function is a set of points, and if there is a computation error, you divert the train from the happy path, from which it can never return.
+- All future functions the error interacts with, never execute any code, the payload continues uninspected through each station on its way to the destination. only then do we need to find out what is inside with .unwrap().
```

Comment:
why .unwrap? in most cases .expect or .map are preferrable as they allow custom messages or handling the Err in some other way.

Reply:
Excellent point, I should mention the various ways to handle it